# NAFuse

NAFuse (Nucleic Acid Fusion) is an integrated methodology that combines and compares the output of a gene fusion detector with that of the BAM-file generated by whole genome sequencing. This allows for the detection of breakpoints in both partner genes/regions on both the RNA and DNA levels without having to manually search output files generated during downstream processing of BAM-files. The main idea of this method is to help to identify, among the huge amount of gene fusions detected with RNA-seq, new gene fusions with DNA-level support.

## Rationale

The BAM file from a sequencing contains information about the reads. When converted in a BED format, the latter contains in each line information about a genomic interval on a specific chromosome with the given start, end positions and read description (Figure 1). 


![NAFuse](https://github.com/user-attachments/assets/26cb9e26-4c2c-42db-886e-73c58785bae5)

**Fig 1. How sequencing reads can align across the genome.**
The two ends of the read can align to (A) the same gene, (B) different genes on the same chromosome, or (C) different genes on different chromosomes. (D) The breakpoint at the DNA-level could be located outside of the partner genes and we therefore increased gene boundaries (start and end position) by 50000 bp to account for such events. https://doi.org/10.1002/path.6219

## Getting Started

NAFuse is implemented in Python and it was tested in a version-controlled conda environment. Before you can run this repository, make sure to install and set up:

- Git. `https://github.com/git-guides/install-git`
- Conda. `Install it through miniconda or anaconda  `
- BEDTools Suite (v. 2.26.0 or later). `https://bedtools.readthedocs.io/en/latest/content/installation.html`
- SamTools Suite. `https://www.htslib.org/download/`

Please make sure to export the path for BEDtools and Samtools to make them directly executble from you directory.

## Project setup

### Installation

Clone the NAFuse repository fro the Github Page and execute the following to generate a conda environment with all you need to run the NAFuse

```
git clone https://github.com//ValeriaDifilippo/NAFuse
cd NAFuse
conda env create -f bin/config/environment.yml
conda activate NAFuse

# Create inputs and results directories in the NAFuse main folder 
mkdir -p input_files 
mkdir -p results
```
or dowload manually the repository from the Git up page and then

```
cd NAFuse
conda env create -f bin/config/environment.yml
conda activate NAFuse

# Create inputs and results directories in the NAFuse main folder 
mkdir -p input_files 
mkdir -p results
```

**Project folder structure**

The project is organized into the following main directories:

- _bin/_
This directory contains all the source codes and scripts used to run NAFuse. In the _config/_ subdirectory you can find the file to set up your environment.

- _annotation/_ 
This directory contains annotation files GRCh37 and GRCh38.112 from Ensemble. Depending on how the chromosome is specified in the bam file you might want to use the *_chr.bed or the *.bed. If you need different files please download them by yourself. 

- _examples/_
This directory contains examples of how the format of the *DNA.bed and *RNA.txt should look like.

- _input_files/_
Contains the files obtained after running the commands in `Step1` and `Step2`. 

- _results/_
This directory stores the output and results generated by running NAFuse.

## Usage

### Step1

This step creates the DNA and RNA files for the NAFuse core.

**DNA**

From the _bin/_ folder run the workflow management system snakemake.This command will convert the bam file into a bed format.

```
snakemake -j 4 --config input_dir=/path/to/BAM/files/ output_dir=/path/to/input_files/ annotation_file=/path/to/annotation/file/*.bed
```
```
-j, --jobs     Specifies how many jobs Snakemake can run in parallel. For example 4 means that Snakemake will utilize 4 CPU cores to execute the workflow faster by running independent tasks in parallel.

--config       Allows you to pass custom configuration variables to the Snakemake workflow. In this case:
                     1. input_dir        This is the directory where your BAM files are located.
                     2. output_dir       This should be the path to the input_files folder, where Snakemake will store the outputs from the workflow and the _RNA.txt files from Step2.
                     3. annotation_file  This is the file containing genomic annotations in BED format in the annotation/ directory.
```
IMPORTANT!!
Make sure that bedtools and samtools are availble and you exported the path. 

> The pipeline:
> 1.	Exctract the reads.
> 2.	Annotate the bed file with the gene name provided from the annotation file.
> 3.	Generate a .bed file containing the read information. where each line show where the two ends of the reads align across the genome, representing the 5' and 3' partners of a gene fusion.
> 4.	Remove the intermediate files generated.

**RNA**

The output from a fusion genes caller needs to be modified in order to have a file with two column representing the gene fusions. 

**5' prime** | **3' prime**
-- | --
RB1 | EWSR1
EWSR1 | KL
KL | CMSS1

Make sure to save this file in the _input_files/_ directory with the same name of the matched DNA but with *RNA.txt. Otherwise NAFuse won't work.

If needed use some filtering steps to remove known false positive gene fusions. For example if you are using fusion catcher (https://github.com/ndaniel/fusioncatcher/tree/master), according to your needs, you can apply the following filters

**Column** | **Filters applied**
-- | --
Spanning unique reads                     | **Keep** `>=3`
Counts of common mapping reads | **Keep** `<1`
Fusion description  | **Remove** `1000genomes, banned, bodymap2, cacg, conjoing, cortex, duplicates, gtex, healthy, hpa, mt, non cancer tissues, non tumor cells, pair pseudo genes, paralogs, rrna, similar symbols, tcga-normal, trna, yrna`

Example code

```
import pandas as pd

sample = pd.read_csv('/path/to/file/',index_col=False, delimiter = '\t',low_memory=False)
remove = sample[sample["Fusion_description"].str.contains("1000genomes|banned|bodymap2|cacg|conjoing|cortex|duplicates|gtex|healthy|hpa|mt|non_cancer_tissues|non_tumor_cells|pair_pseudo_genes|paralogs|rrna|similar_symbols|tcga-normal|trna|yrna") == False]
keep = remove.loc[remove["Spanning_unique_reads"] >= 3 ]
keep1 = keep.loc[keep["Counts_of_common_mapping_reads"] < 1 ]
first_two_columns = keep1.iloc[:, :2]
df_cleaned = first_two_columns.drop_duplicates()
df_sorted = df_cleaned.sort_values(by=df_cleaned.columns[0])
df_sorted.to_csv('/path/to/inputs/'+i+'.txt', index = False, header=False, sep='\t')
```
**IMPORTANT!!**
In the _input_files/_ each pair of DNA and RNA needs to have matching names, like

> sample1_DNA.bed
> sample1_RNA.txt

You can check the correct formats in the _example/_ directory.

### Step2

Run NAFuse core 
```
python NAFuse.py -dir /path/to/input_files -o /path/to/results
```
```
python NAFuse.py --help

usage: NAFuse.py [-h] -dir DIR -o O

NAFuse allows for the detection of breakpoints in both partner genes/regions on both the RNA and DNA levels without having to manually search output files generated during downstream processing of BAM-files.

optional arguments:
  -h, --help  show this help message and exit
  -dir, DIR    Directory containing the DNA and RNA files
  -o,O        Output directory for gene fusions results
```

The final outputs of NAFuse will be saved in the _results/_ directory. The tool will produce two different outputs for each sample

`sample1_validated_gene_fusion.txt`
A file containing the list of the validated gene fusions resulting from the matches between RNA and DNA.
`sample1_supporting_reads.txt`
A File containing the the validated gene fusions with the support of the reads information.


### Validation and Tips

-	NAFuse was tested in cases reported in https://doi.org/10.1016/j.labinv.2023.100283 and https://doi.org/10.1002/path.6219. It was able to successfully detect all the gene fusions described when DNA and RNA were available, such as _FOSL1::RELA_, _FUS::TFCP2_ and the _TP53_ fusions.
-	In some instances, the breakpoint on the DNA level could be located upstream of the 3â€™ partner gene. To solve this, we increased gene boundaries (start and stop) by 50 000 bp each.
-	Use the correct gene annotation file. Be sure to use the same gene annotation used by your gene fusion detector of choice. It can be necessary to merge multiple gene annotations files.

### Citations

Please if you use NAFuse, cite:
Valeria Difilippo, Karim H. Saba, Emelie Styring, Linda Magnusson, Jenny Nilsson, Michaela Nathrath, Daniel Baumhoer, Karolin H. Nord, Osteosarcomas With Few Chromosomal Alterations or Adult Onset Are Genetically Heterogeneous, Laboratory Investigation, Volume 104, Issue 1, 2024, 100283, ISSN 0023-6837, https://doi.org/10.1016/j.labinv.2023.100283.

